cns/deviceplugin/pluginmanager.go

```diff
@@ -0,0 +1,147 @@
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-container-networking/crd/multitenancy/api/v1alpha1"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+var ErrUnknownDeviceType = errors.New("unknown device type")
+
+const (
+	defaultDevicePluginDirectory = "/var/lib/kubelet/device-plugins"
+	defaultDeviceCheckInterval   = 5 * time.Second
+)
+
+type pluginManagerOptions struct {
+	devicePluginDirectory string
+	kubeletSocket         string
+	deviceCheckInterval   time.Duration
+}
+
+type pluginManagerOption func(*pluginManagerOptions)
+
+func PluginManagerSocketPrefix(prefix string) func(*pluginManagerOptions) {
+	return func(opts *pluginManagerOptions) {
+		opts.devicePluginDirectory = prefix
+	}
+}
+
+func PluginManagerKubeletSocket(socket string) func(*pluginManagerOptions) {
+	return func(opts *pluginManagerOptions) {
+		opts.kubeletSocket = socket
+	}
+}
+
+func PluginDeviceCheckInterval(i time.Duration) func(*pluginManagerOptions) {
+	return func(opts *pluginManagerOptions) {
+		opts.deviceCheckInterval = i
+	}
+}
+
+// PluginManager runs device plugins for vnet nics and ib nics
+type PluginManager struct {
+	Logger        *zap.Logger
+	VnetNICPlugin *Plugin
+	IBNICPlugin   *Plugin
+	options       pluginManagerOptions
+}
+
+func NewPluginManager(l *zap.Logger, initialVnetNICCount, initialIBNICCount int, opts ...pluginManagerOption) *PluginManager {
+	logger := l.With(zap.String("component", "devicePlugin"))
+	socketWatcher := NewSocketWatcher(logger)
+	options := pluginManagerOptions{
+		devicePluginDirectory: defaultDevicePluginDirectory,
+		kubeletSocket:         v1beta1.KubeletSocket,
+		deviceCheckInterval:   defaultDeviceCheckInterval,
+	}
+	for _, o := range opts {
+		o(&options)
+	}
+	return &PluginManager{
+		Logger: logger,
+		VnetNICPlugin: NewPlugin(logger, string(v1alpha1.DeviceTypeVnetNIC), socketWatcher,
+			getSocketName(options.devicePluginDirectory, v1alpha1.DeviceTypeVnetNIC), initialVnetNICCount,
+			options.kubeletSocket, options.deviceCheckInterval),
+		IBNICPlugin: NewPlugin(logger, string(v1alpha1.DeviceTypeInfiniBandNIC), socketWatcher,
+			getSocketName(options.devicePluginDirectory, v1alpha1.DeviceTypeInfiniBandNIC), initialIBNICCount,
+			options.kubeletSocket, options.deviceCheckInterval),
```

Comment:
it should either block on the pluginmanager being unstarted, or have a mutex on the plugins list to prevent races while allowing dynamic plugin addition